ytypes/leafref_test.go

```diff
@@ -0,0 +1,432 @@
+		},
+		{
+			desc: "blank",
+			in:   "a//b",
+			want: []string{"a", "", "b"},
+		},
+		{
+			desc: "lead trail slash",
+			in:   "/a/b/c/",
+			want: []string{"", "a", "b", "c", ""},
+		},
+		{
+			desc: "escape slash",
+			in:   `a/\/b/c`,
+			want: []string{"a", `\/b`, "c"},
+		},
+		{
+			desc: "internal key slashes",
+			in:   `a/b[key1 = ../x/y key2 = "z"]/c`,
+			want: []string{"a", `b[key1 = ../x/y key2 = "z"]`, "c"},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			if got, want := splitPath(tt.in), tt.want; !reflect.DeepEqual(got, want) {
+				t.Errorf("%s: got: %v, want: %v", tt.desc, got, want)
+			}
+		})
+	}
+}
+
+func TestSplitUnescaped(t *testing.T) {
+	tests := []struct {
+		desc string
+		in   string
+		want []string
+	}{
+		{
+			desc: "simple",
+			in:   "a/b/c",
+			want: []string{"a", "b", "c"},
+		},
+		{
+			desc: "blank",
+			in:   "a//b",
+			want: []string{"a", "", "b"},
+		},
+		{
+			desc: "lead trail slash",
+			in:   "/a/b/c/",
+			want: []string{"", "a", "b", "c", ""},
+		},
+		{
+			desc: "escape slash",
+			in:   `a/\/b/c`,
+			want: []string{"a", `\/b`, "c"},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			if got, want := splitUnescaped(tt.in, '/'), tt.want; !reflect.DeepEqual(got, want) {
+				t.Errorf("%s: got: %v, want: %v", tt.desc, got, want)
+			}
+		})
+	}
+}
+
+func TestSplitUnquoted(t *testing.T) {
+	tests := []struct {
+		desc     string
+		in       string
+		splitStr string
+		want     []string
+	}{
+		{
+			desc:     "simple",
+			in:       "a/b/c",
+			splitStr: "/",
+			want:     []string{"a", "b", "c"},
+		},
+		{
+			desc:     "blank",
+			in:       "a//b",
+			splitStr: "/",
+			want:     []string{"a", "", "b"},
+		},
+		{
+			desc:     "lead trail slash",
+			in:       "/a/b/c/",
+			splitStr: "/",
+			want:     []string{"", "a", "b", "c", ""},
+		},
+		{
+			desc:     "quoted",
+			in:       `a/"/"b"/"/c`,
+			splitStr: "/",
+			want:     []string{"a", `"/"b"/"`, "c"},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			if got, want := splitUnquoted(tt.in, tt.splitStr), tt.want; !reflect.DeepEqual(got, want) {
+				t.Errorf("%s: got: %v, want: %v", tt.desc, got, want)
+			}
+		})
+	}
+}
+
+func TestExtractKeyValue(t *testing.T) {
+	tests := []struct {
+		desc       string
+		in         string
+		wantErr    string
+		wantPrefix string
+		wantKey    string
+		wantValue  string
+	}{
+		{
+			desc:       "literal",
+			in:         `b[key = "value"]`,
+			wantPrefix: "b",
+			wantKey:    "key",
+			wantValue:  `"value"`,
+		},
+		{
+			desc:       "spacing",
+			in:         `b[key="value"]`,
+			wantPrefix: "b",
+			wantKey:    "key",
+			wantValue:  `"value"`,
+		},
+		{
+			desc:       "quotes",
+			in:         `b[key="[=value=]"]`,
+			wantPrefix: "b",
+			wantKey:    "key",
+			wantValue:  `"[=value=]"`,
+		},
+		{
+			desc:       "path",
+			in:         "b[key = ../a/b/c]",
```

Comment:
Unquoted keys aren't valid XPATH - so wouldn't ever be seen here, IMHO, this should return an error.

Reply:
Done.